src/devices/MatrixKeyboard/MatrixKeyboard.cs

```diff
@@ -0,0 +1,147 @@
+        /// </summary>
+        public double ScanFreq
+        {
+            get => 1000 / (scanTimer.Interval / colPins.Length);
+            set => scanTimer.Interval = 1000 / (value * rowPins.Length);
+        }
+
+        /// <summary>
+        /// Initialize keyboard
+        /// </summary>
+        /// <param name="gpioController">GPIO controller</param>
+        /// <param name="rowPins">Row pins</param>
+        /// <param name="colPins">Column pins</param>
+        /// <param name="scanFreq">Scanning frequency</param>
+        public MatrixKeyboard(IGpioController gpioController, IEnumerable<int> rowPins, IEnumerable<int> colPins, double scanFreq = 50)
+        {
+            gpio = gpioController;
+
+            this.rowPins = rowPins.ToArray();
+            this.colPins = colPins.ToArray();
+            buttonValues = new PinValue[this.rowPins.Length * this.colPins.Length];
+
+            scanTimer = new Timer(1000 / (scanFreq * this.rowPins.Length));
+            scanTimer.Elapsed += ScanTimerElapsed;
+        }
+
+        /// <summary>
+        /// Open GPIO pins then start keyboard scanning
+        /// </summary>
+        public void StartScan()
+        {
+            for (var i = 0; i < rowPins.Length; i++)
+            {
+                gpio.OpenPin(rowPins[i], PinMode.Output);
+            }
+            for (var i = 0; i < colPins.Length; i++)
+            {
+                gpio.OpenPin(colPins[i], PinMode.Input);
+            }
+            scanTimer.Start();
+        }
+
+        /// <summary>
+        /// Stop keyboard scanning then close GPIO pins
+        /// </summary>
+        public void StopScan()
+        {
+            scanTimer.Stop();
+            for (var i = 0; i < rowPins.Length; i++)
+            {
+                gpio.ClosePin(rowPins[i]);
+            }
+            for (var i = 0; i < colPins.Length; i++)
+            {
+                gpio.ClosePin(colPins[i]);
+            }
+        }
+
+        /// <summary>
+        /// Get all buttons' values
+        /// </summary>
+        public ReadOnlySpan<PinValue> Values => buttonValues.AsSpan();
+
+        /// <summary>
+        /// Get buttons' values by row
+        /// </summary>
+        /// <param name="row">Row index</param>
+        public ReadOnlySpan<PinValue> RowValues(int row) => buttonValues.AsSpan(row * rowPins.Length, colPins.Length);
+
+        /// <summary>
+        /// Keyboard event
+        /// </summary>
+        public event PinChangeEventHandler PinChangeEvent;
+
+        /// <summary>
+        /// Delegate of keyboard event
+        /// </summary>
+        public delegate void PinChangeEventHandler(object sender, MatrixKeyboardEventArgs pinValueChangedEventArgs);
+
+        public void Dispose()
+        {
+            scanTimer?.Dispose();
+            scanTimer = null;
+
+            gpio?.Dispose();
+            gpio = null;
+        }
+
+        private readonly int[] rowPins;
```

Comment:
you should move all these privates to the top of the class just so that the class is more readable. Also, naming doesn't seem to follow conventions, for example, private fields should be named _rowPins. you can find more info here: [https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md)